### PR TITLE
[WIP] Frontend forms / Login

### PR DESCRIPTION
**Pull request by phile314**
I have been playing a bit around with the elm code. When I tried to get the login form working, I had to actually make some quite big changes as I didn't want to put all ``Msg`` constructors into the top-level ``Msg``. I mostly followed the example from https://github.com/rtfeldman/elm-spa-example to get it working.

This is a work-in-progress, but it would be good to discuss what your plans for the hercules frontend and hercules itself are?

### CODE REVIEW

**Comment by expipiplus1** [comment]
@domenkozar's the person for the frontend stuff!

File: frontend/src/Models.elm [hunk 1]
```diff
@@ -96,6 +96,15 @@ type AjaxError msg
...
 
 
+{-| Main type representing model of current page
+-}
+type Page
+    = HomePage
+    | LoginPage Login.Model
+    | ProjectPage String
+    | NewProjectPage
+    | JobsetPage2 String String
```
Comment (domenkozar) [inline frontend/src/Models.elm]: Why not just `JobsetPage`? :)

File: frontend/src/Update.elm [hunk 2]
```diff
@@ -1,74 +1,78 @@
...
             ( model, Cmd.none )
 
-        UrlChange location ->
-            let
-                page = Maybe.withDefault Home (parsePath pageParser location)
-            in
-            ( { model | currentPage = page }
-            , title (pageToTitle page)
-            )
+        ( _, _ ) ->
```
Comment (domenkozar) [inline frontend/src/Update.elm]: I'd prefer not to have catch-all, to actually error out if something is missing.

**Review by domenkozar — CHANGES_REQUESTED** [review]
Looks great progress! Do you need some help going forward?

**Comment by phile314-fh** [comment]
So the big architectural change is really with the new ``Msg`` with the sub-msg types inside the page. Also the new ``Page`` type and the corresponding changes in the routing/update functions. Do you think this is going in the right direction? (This stuff is quite painful to change, so that why I am bringing this up now even though this PR is still work-in-progress.)

I think I will finish up the Login page, with the goal to merge this as a first step. Second part probably would be to implement e.g. the Jobset display with the actual REST API.

**Comment by domenkozar** [comment]
@phile314 it does go in the right direction. Ideally, each component would be self-contained. And then routing functions can work on any component independently. Good work!

File: frontend/src/Update.elm (hunk 2 above)
Comment (phile314) [reply frontend/src/Update.elm]: Hm, I tried doing that but then the compiler complains that there are unhandled cases:
```
    (Msg.LoginMsg _, Models.HomePage)
    (Msg.LoginMsg _, Models.ProjectPage _)
    (Msg.LoginMsg _, Models.NewProjectPage)
    (Msg.LoginMsg _, Models.JobsetPage _ _)
```
This are basically impossible cases, but the compiler doesn't see that. What about putting a Debug.crash inside the catch-all instead?

**Pull request closed without merging**